src/Microsoft.Azure.ServiceBus/ServiceBusDiagnosticsSource.cs

```diff
@@ -0,0 +1,773 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Azure.ServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Threading.Tasks;
+
+    using Microsoft.Azure.ServiceBus.Diagnostics;
+
+    internal class ServiceBusDiagnosticSource
+    {
+        public const string DiagnosticListenerName = "Microsoft.Azure.ServiceBus";
+        public const string BaseActivityName = "Microsoft.Azure.ServiceBus.";
+
+        public const string ExceptionEventName = BaseActivityName + "Exception";
+        public const string ProcessActivityName =  BaseActivityName + "Process";
+
+        public const string ActivityIdPropertyName = "Diagnostic-Id";
+        public const string CorrelationContextPropertyName = "Correlation-Context";
+        public const string RelatedToTag = "RelatedTo";
+        public const string MessageIdTag = "MessageId";
+        public const string SessionIdTag = "SessionId";
+
+        private static readonly DiagnosticListener DiagnosticListener = new DiagnosticListener(DiagnosticListenerName);
+        private readonly string entityPath;
+        private readonly Uri endpoint;
+
+        public ServiceBusDiagnosticSource(string entityPath, Uri endpoint)
+        {
+            this.entityPath = entityPath;
+            this.endpoint = endpoint;
+        }
+
+        public static bool IsEnabled()
```

Comment:
Suggestion: Make this a property. i.e.
```C#
public static bool IsEnabled => DiagnosticListener.IsEnabled();
```

I think it reads more naturally at the call-site.

Reply:
it also creates a false impression on the caller side, that `ServiceBusDiagnosticSource.IsEnabled` value is static and could be called every time instead of caching the value (like it done now)

```C#
            bool isDiagnosticsEnabled = ServiceBusDiagnosticSource.IsEnabled();
            Activity activity = isDiagnosticsEnabled ? this.diagnosticSource.ReceiveDeferredStart(sequenceNumberList) : null;
            try
            {
                ...
            }
            catch (Exception exception)
            {
                if (isDiagnosticsEnabled)
                {
                    this.diagnosticSource.ReportException(exception);
                }
            }
```

with property it will be more natural to write
```C#
            Activity activity = ServiceBusDiagnosticSource.IsEnabled? this.diagnosticSource.ReceiveDeferredStart(sequenceNumberList) : null;
            try
            {
                ...
            }
            catch (Exception exception)
            {
                if (ServiceBusDiagnosticSource.IsEnabled)
                {
                    this.diagnosticSource.ReportException(exception);
                }
            }
```

While it's quite cheap to call `DiagnosticListener.IsEnabled()`,  I still prefer result to be cached for particular operations not only because of performance, but also consistency